benchmark-base/src/main/java/org/conscrypt/ClientSocketBenchmark.java

```diff
@@ -0,0 +1,156 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Benchmark for comparing performance of client socket implementations. All benchmarks use Netty
+ * with tcnative as the server.
+ */
+public final class ClientSocketBenchmark {
+    /**
+     * Provider for the benchmark configuration
+     */
+    interface Config {
+        SocketType socketType();
+        int messageSize();
+        String cipher();
+        WrappedSocketType wrappedSocketType();
+    }
+
+    private ClientEndpoint client;
+    private ServerEndpoint server;
+    private byte[] message;
+    private ExecutorService executor;
+    private Future<?> sendingFuture;
+    private volatile boolean stopping;
+
+    private static final AtomicLong bytesCounter = new AtomicLong();
+    private AtomicBoolean recording = new AtomicBoolean();
+
+    ClientSocketBenchmark(Config config) throws Exception {
+        recording.set(false);
+
+        message = newTextMessage(config.messageSize());
+
+        // Always use the same server for consistency across the benchmarks.
+        server = SocketType.CONSCRYPT_ENGINE.newServer(
+                WrappedSocketType.CHANNEL, config.messageSize(), getProtocols(), ciphers(config));
+
+        server.setMessageProcessor(new ServerEndpoint.MessageProcessor() {
+            @Override
+            public void processMessage(byte[] inMessage, int numBytes, OutputStream os) {
+                if (recording.get()) {
+                    // Server received a message, increment the count.
+                    bytesCounter.addAndGet(numBytes);
+                }
+            }
+        });
+        Future<?> connectedFuture = server.start();
+
+        client = config.socketType().newClient(
+            config.wrappedSocketType(), server.port(), getProtocols(), ciphers(config));
+        client.start();
+
+        // Wait for the initial connection to complete.
+        connectedFuture.get(5, TimeUnit.SECONDS);
+
+        executor = Executors.newSingleThreadExecutor();
+        sendingFuture = executor.submit(new Runnable() {
+            @Override
+            public void run() {
+                Thread thread = Thread.currentThread();
+                while (!stopping && !thread.isInterrupted()) {
+                    client.sendMessage(message);
+                }
+            }
+        });
+    }
+
+    void close() throws Exception {
+        stopping = true;
+        client.stop();
+        server.stop();
+        executor.shutdown();
+        executor.awaitTermination(5, TimeUnit.SECONDS);
+        sendingFuture.get(5, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Simple benchmark for throughput.
+     */
+    void throughput() throws Exception {
+        recording.set(true);
+        // No need to do anything, just sleep here.
+        Thread.sleep(1001);
```

Comment:
I actually meant that setting up the benchmark as "how many messages can I send in a second" is going to give us imprecise results, since Thread.sleep() won't necessarily sleep a consistent amount of time, so if instead we benchmarked how long it takes to send a fixed number of messages, that would give us a more reliable result.  Doesn't necessarily need to be changed in this change, but worth considering for the future.